reduce_confusion_matrix.py

```diff
@@ -0,0 +1,116 @@
+# Labelled Phrase,Detected Phrase,Intent Id,Intent Name,Top Match Intent Id,Top Match Intent Name,Top Match Score,Entropy,Uncertainty,Margin Score,Result Type
+
+@click.command()
+@click.option('-m', '--top_mispredictions', type=int, default=5, help='Number of top mispredictions')
+@click.option('-f', '--phrases_filename', type=str, default='./data/phrases.csv', help='Phrases')
+def main(phrases_filename: str, top_mispredictions: int) -> None:
+    '''Main function'''
+
+    reduce_confusion_matrix(phrases_filename, top_mispredictions)
+
+def reduce_confusion_matrix(phrases_filename: str, top_mispredictions: int) -> None:
+    '''Reduces Confusion matrix'''
+    df = pandas.read_csv(phrases_filename)
+    
+    # summarise
+    print('Summary of df')
+    print(df[["Intent Name","Result Type"]].groupby(["Result Type"]).count())
+
+    # create confusion matrix
+    labels = sorted(list(set(df["Intent Name"])))
+    print(f"Total Number of intents: {len(labels)}")
+    matrix = confusion_matrix(df["Intent Name"],df["Top Match Intent Name"],labels = labels)
+    
+    # replace true positives to 0 and sum all the incorrect predictions
+    total_mispredictions = 0
+    for i in range(len(matrix)):
+        for j in range(len(matrix[i])):
+            if i == j:
+                matrix[i][j] = 0
+            else:
+                total_mispredictions = total_mispredictions + matrix[i][j]
+    print("\nTrue positives are replaced with 0\n")
+    
+    # confusion matrix into dataframe
+    df_matrix = pandas.DataFrame(data=matrix,index=labels,columns=labels)
+
+    # get the list of top mispredictions e.g. top 10 largest number of incorrect predictions
+    top_mispredictions_list = (heapq.nlargest(top_mispredictions,matrix.max(1)))
+
+    # determining the row and col name of those top x incorrect predictions
+    intents_col = set()
+    intents_row = set()
+    for row in df_matrix.iterrows():
+        for index in row[1].index:
+            if row[1][index] in top_mispredictions_list:
+                intents_col.add(index)
+                intents_row.add(row[0])
+
+    # create reduced matrix
+    reduced_matrix = df_matrix.loc[list(intents_row)][list(intents_col)]
+    reduced_matrix.loc["Total"] = reduced_matrix.sum(axis = 0)
+    reduced_matrix = reduced_matrix.sort_values(by=["Total"],axis=1,ascending=False)
+    reduced_matrix["Total"] = reduced_matrix.sum(axis = 1)
+    reduced_matrix.sort_values(by=["Total"],axis=0,inplace=True,ascending=False)
+    
+    # reorder Total row as last row
+    reorder_list = list(reduced_matrix.index)
+    del reorder_list[0]
+    reorder_list.append("Total")
+    reduced_matrix = reduced_matrix.reindex(reorder_list)
+
+    print("Reduced Confusion Matrix")
+    print(reduced_matrix)
+
+    print(f"\nPercentage of confusions represented by the matrix: {round((reduced_matrix.loc['Total']['Total']/total_mispredictions)*100,2)} %")
+
+    # determine top intent pair the model is getting confused
```

Comment:
This then as a separate summary function will be clearer the type of reduced_matrix it then accepts.